runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -1573,16 +1915,69 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
 
    TR::TreeTop* directAccessTreeTop =
       genDirectAccessCodeForUnsafeGetPut(callNodeTreeTop->getNode(), false, true);
-   TR::TreeTop* arrayDirectAccessTreeTop = conversionNeeded
-      ? genDirectAccessCodeForUnsafeGetPut(callNodeWithConversion, conversionNeeded, true)
-      : NULL;
+
+   // We need to generate an arrayDirectAccessBlock AND a directAccessBlock in the following cases:
+   // 1.) arrayTestNeeded == true: arrayDirectAccessBlock used for array access,
+   //                               directAccessBlock used for non-array/non-class object access
+   // 2.) !arrayTestNeeded && arrayBlockNeeded: arrayDirectAccessBlock used for array access,
+   //                                            directAccessBlock used if object is NULL
+   // Otherwise, only the directAccessBlock is needed.
+   TR::TreeTop* arrayDirectAccessTreeTop;
+
+   if (conversionNeeded && (arrayTestNeeded || arrayBlockNeeded))
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(callNodeWithConversion, conversionNeeded, true);
+   else if (arrayTestNeeded || arrayBlockNeeded)
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(callNodeTreeTop->getNode(), false, true);
```

Comment:
As with `TR_J9InlinerPolicy::createUnsafePutWithOffset`, this might result in a block of IL that is identical to that produced for `directAccessTreeTop` above.